x-pack/filebeat/input/gcs/job/job.go

```diff
@@ -0,0 +1,179 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+//go:build !aix
+// +build !aix
+
+package job
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	"cloud.google.com/go/storage"
+
+	cursor "github.com/elastic/beats/v7/filebeat/input/v2/input-cursor"
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/x-pack/filebeat/input/gcs/state"
+	"github.com/elastic/beats/v7/x-pack/filebeat/input/gcs/types"
+	"github.com/elastic/elastic-agent-libs/mapstr"
+)
+
+type Job interface {
+	Do(ctx context.Context, id string) error
+	Name() string
+	Timestamp() *time.Time
+	Source() *types.Source
+}
+
+type GcsInputJob struct {
+	bucket    *storage.BucketHandle
+	object    *storage.ObjectAttrs
+	objectURI string
+	state     *state.State
+	src       *types.Source
+	publisher cursor.Publisher
+}
+
+// NewGcsInputJob, returns an instance of a job , which is a unit of work that can be assigned to a go routine
+func NewGcsInputJob(bucket *storage.BucketHandle, object *storage.ObjectAttrs, objectURI string,
+	state *state.State, src *types.Source, publisher cursor.Publisher,
+) Job {
+	return &GcsInputJob{
+		bucket:    bucket,
+		object:    object,
+		objectURI: objectURI,
+		state:     state,
+		src:       src,
+		publisher: publisher,
+	}
+}
+
+func (gcsij *GcsInputJob) Do(ctx context.Context, id string) error {
+	var fields mapstr.M
+
+	if types.AllowedContentTypes[gcsij.object.ContentType] {
+		data, err := gcsij.extractData(ctx)
+		if err != nil {
+			return fmt.Errorf("job with jobId %s encountered an error : %w", id, err)
+		}
+
+		reader := io.NopCloser(bytes.NewReader(data.Bytes()))
+		defer func() {
+			err = reader.Close()
+			if err != nil {
+				err = fmt.Errorf("failed to close json reader with error : %w", err)
+			}
```

Comment:
This is an ineffectual assignment. If you want this to be returned, you will need a named error return in the signature.